rcl_logging_interface/src/logging_dir.c

```diff
@@ -0,0 +1,95 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <rcutils/allocator.h>
+#include <rcutils/error_handling.h>
+#include <rcutils/filesystem.h>
+#include <rcutils/get_env.h>
+#include <rcutils/strdup.h>
+
+#include "rcl_logging_interface/rcl_logging_interface.h"
+
+rcl_logging_ret_t
+rcl_logging_get_logging_directory(rcutils_allocator_t allocator, char ** directory)
+{
+  if (NULL == directory) {
+    RCUTILS_SET_ERROR_MSG("directory argument must not be null");
+    return RCL_LOGGING_RET_INVALID_ARGUMENT;
+  }
+  if (NULL != *directory) {
+    RCUTILS_SET_ERROR_MSG("directory argument must point to null");
+    return RCL_LOGGING_RET_INVALID_ARGUMENT;
+  }
+
+  const char * log_dir_env;
+  const char * err = rcutils_get_env("ROS_LOG_DIR", &log_dir_env);
+  if (NULL != err) {
+    RCUTILS_SET_ERROR_MSG("rcutils_get_env failed");
+    return RCL_LOGGING_RET_ERROR;
+  }
+  if ('\0' != *log_dir_env) {
+    *directory = rcutils_strdup(log_dir_env, allocator);
+    if (NULL == *directory) {
+      RCUTILS_SET_ERROR_MSG("rcutils_strdup failed");
+      return RCL_LOGGING_RET_ERROR;
+    }
+  } else {
+    const char * ros_home_dir_env;
+    err = rcutils_get_env("ROS_HOME", &ros_home_dir_env);
+    if (NULL != err) {
+      RCUTILS_SET_ERROR_MSG("rcutils_get_env failed");
+      return RCL_LOGGING_RET_ERROR;
+    }
+    if ('\0' == *ros_home_dir_env) {
+      ros_home_dir_env = rcutils_join_path("~", ".ros", allocator);
+      if (NULL == ros_home_dir_env) {
+        RCUTILS_SET_ERROR_MSG("rcutils_join_path failed");
+        return RCL_LOGGING_RET_ERROR;
+      }
+    }
+    *directory = rcutils_join_path(ros_home_dir_env, "log", allocator);
+    if (NULL == *directory) {
+      RCUTILS_SET_ERROR_MSG("rcutils_join_path failed");
+      return RCL_LOGGING_RET_ERROR;
+    }
+  }
+
+  // Expand home directory
+  if ('~' == (*directory)[0]) {
+    const char * homedir = rcutils_get_home_dir();
+    if (NULL == homedir) {
+      RCUTILS_SET_ERROR_MSG("failed to get the home directory");
+      return RCL_LOGGING_RET_ERROR;
+    }
+    if (1 == strlen(*directory)) {
+      allocator.deallocate(*directory, allocator.state);
+      *directory = rcutils_strdup(homedir, allocator);
+      if (NULL == *directory) {
+        RCUTILS_SET_ERROR_MSG("rcutils_strdup failed");
+        return RCL_LOGGING_RET_ERROR;
+      }
+    } else {
+      char * directory_not_expanded = *directory;
+      // Try to avoid having a double slash
+      const int offset = (2 <= strlen(*directory) && '/' == directory_not_expanded[1]) ? 2 : 1;
+      *directory = rcutils_join_path(homedir, directory_not_expanded + offset, allocator);
```

Comment:
Last nit, I wouldn't handle double `/`.
This snippet as-is doesn't work on Windows, and in all OS I know of double path separators are automatically simplified (e.g. `/my/home//log` works ok on Linux/macOS and `C:\my\home\\log` also works on Windows).
```suggestion
      *directory = rcutils_join_path(homedir, directory_not_expanded + 1, allocator);
```

If you want to handle this case, you would need something like https://github.com/ros2/rcutils/blob/22d5e564da8b0eae9d8c786a8e4cd78dceccb4f5/src/filesystem.c#L38-L42 (which should be moved to `filesystem.h`.

Reply:
the main reason why I did that offset thing is because it was outputing double slashes every time otherwise. I'll try to figure out why exactly, but in the meantime I've opened https://github.com/ros2/rcutils/pull/294

Reply:
> the main reason why I did that offset thing is because it was outputing double slashes every time otherwise

Ah, I imagine that was affecting the tests.
The issue is that the tests will fail on Windows with the current code.

You actually don't need to "join paths", you just need to join strings in this case.
Something that the following would fix the issue:
```c
size_t homedir_len = strlen(homedir);
size_t directory_not_expanded_len = strlen(directory_not_expanded);

char * new_directory = malloc(directory_not_expanded + homedir_len + 1);
// ERROR HANDLING HERE
strncpy(new_directory, homedir, homedir_len);
strncpy(new_directory + homedir_len, directory_not_expanded_len, directory_not_expanded_len + 1);
```
And you don't even need to handle the `strlen(directory_not_expanded) == 1u` case separately.
It would be better to have a join strings function in `rcutils`, but adding another blocker to this sounds unnecesary.

Reply:
> You actually don't need to "join paths", you just need to join strings in this case.

Reply:
> You actually don't need to "join paths", you just need to join strings in this case.

thank you for pointing it out

> It would be better to have a join strings function in rcutils

we do have `rcutils_format_string` :wink: and it can use our allocator, so I'll use that if you don't mind

Reply:
PTAL at 93ad695580b76cddf3c510add58685e5e97e61e8